src/distilabel/steps/deita.py

```diff
@@ -0,0 +1,192 @@
+            Defaults to `0.9`.
+        normalize_embeddings: Whether to normalize the embeddings before computing the cosine
+            distance. Defaults to `True`.
+
+    Runtime parameters:
+        - `data_budget`: The desired size of the dataset after filtering.
+        - `diversity_threshold`: If a row has a cosine distance with respect to it's nearest
+            neighbor greater than this value, it will be included in the filtered dataset.
+
+    Input columns:
+        - evol_instruction_score (`float`): The score of the instruction generated by
+            `ComplexityScorer` step.
+        - evol_response_score (`float`): The score of the response generated by
+            `QualityScorer` step.
+        - embedding (`List[float]`): The embedding generated for the conversation of the
+            instruction-response pair using `GenerateEmbeddings` step.
+
+    Output columns:
+        - deita_score (`float`): The DEITA score for the instruction-response pair.
+        - nearest_neighbor_distance (`float`): The cosine distance between the embeddings
+            of the instruction-response pair.
+
+    Reference:
+        - [`What Makes Good Data for Alignment? A Comprehensive Study of Automatic Data Selection in Instruction Tuning`](https://arxiv.org/abs/2312.15685)
+    """
+
+    data_budget: RuntimeParameter[int] = Field(
+        default=None, description="The desired size of the dataset after filtering."
+    )
+    diversity_threshold: RuntimeParameter[float] = Field(
+        default=0.9,
+        description="If a row has a cosine distance with respect to it's nearest neighbor"
+        " greater than this value, it will be included in the filtered dataset.",
+    )
+    normalize_embeddings: RuntimeParameter[bool] = Field(
+        default=True,
+        description="Whether to normalize the embeddings before computing the cosine distance.",
+    )
+
+    _NearestNeighbors: Type["NearestNeighbors"] = PrivateAttr(...)
+
+    def load(self) -> None:
+        try:
+            from sklearn.neighbors import NearestNeighbors
+        except ImportError as ie:
+            raise ImportError(
+                "`scikit-learn` is not installed. Please install it using `pip install huggingface-hub`."
+            ) from ie
+
+        self._NearestNeighbors = NearestNeighbors
+
+    @property
+    def inputs(self) -> List[str]:
+        return ["evol_instruction_score", "evol_response_score", "embedding"]
+
+    @property
+    def outputs(self) -> List[str]:
+        return ["deita_score", "nearest_neighbor_distance"]
+
+    def _compute_deita_score(self, inputs: StepInput) -> StepInput:
+        """Computes the DEITA score for each instruction-response pair. The DEITA score is
+        the product of the instruction score and the response score.
+
+        Args:
+            inputs: The input data.
+
+        Returns:
+            The input data with the DEITA score computed.
+        """
+        for input_ in inputs:
```

Comment:
Not sure about returning the score if the computation was faulty. people might assume their response and instruction where good, while in practice only either one of them might have been correct.

Reply:
I've added the `deita_score_computed_with` which indicates which scores were used to compute `deita_score`